# End-to-end demonstration 

A paper that has now been submitted to Science Advances and is in BioRxiv illustrates the end-to-end process.

The [Impact of Sex on Alternative Splicing](https://www.biorxiv.org/content/10.1101/490904v2), uses data from dbGaP, the Gene Tissue Expression Data set which is controlled access.

The code for this paper is on [GitHub](https://github.com/TheJacksonLaboratory/sbas)

Get access to the Genotype-Tissue Expression [GTEx](https://www.gtexportal.org/home/) RNAseq data (an application to dbGAP for access to the dataset phs000424.v8.v2 is required)

Align each RNAseq sample using hisat2 and create a matrix of counts for each of a variety of splicing types was generated by the rMATS. Specifically, rMATS was run as a [nextflow script](https://github.com/lifebit-ai/rmats-nf/). The script may be modified to run on any platform, the results from this study was performed on the cloudOS/lifebit platform.

Run the Jupyter notebooks from this repository to perform the individual analyses.
This repository documents the interactive analysis for the results of running the rmats-nf pipeline.


