<p align="center">
  <img src="https://static.wixstatic.com/media/b34882_b300f090387248599125482b6750206c.jpg"  width="250" align="right" >
</p>


# Dry bench skills for Researchers - Mini-Courses

![](assets/header.png)

## Mini courses
Today we begin our bi-weekly series, focusing on little bites to introduce you to topics that facilitate your research by providing support and skills to help you create reproducible work, both in workflows and in your analyses, at scale and in the cloud, as well as show how you can do this work .  Every two weeks, beginning today, we will tackle a new topic.

Thursday, March 25 - **Short course on how to perform germline or somatic variant calling with nf-core/sarek** ([Agenda](1_variant_calling/README.md)) 
Thursday, April 8  - **Discovering SRA data and running an End-to-End RNAseq Workflow** ([Agenda](2_sra_and_rnaseq/README.md))

## INSTRUCTORS
- [Anne Deslattes Mays](mailto:adeslat@scitechcon.org), Ph.D., Science and Technology Consulting LLC
- [Christina Chatzipantsiou](mailto:christina@lifebit.ai), Lifebit

## ABOUT

These hands-on mini courses will serve as an introduction to reproducible workflows using Nextflow based pipelines for NGS analyses. The expected outcome of the short course is for the attendees to familiarise with the variant calling workflow, and be able to run locally, on platforms with SLURM executors and on JAX CloudOS on Google Cloud Platform an end-to-end analyses, started from FASTQ and getting output VCF files. We assume that the attendees have basic minimal familiarity with bash scripting and git. 

## PRE-ATTENDENCE READING and ACTIVITIES

If you don't have a GitHub account, please create one here:
[Git Sign up - create an account](https://github.com/join?ref_cta=Sign+up&ref_loc=header+logged+out&ref_page=%2F&source=header-home)

And email your newly created username to adeslat@scitechcon.org and we will create a single mail to add all attendees to the https://github.com/TheJacksonLaboratory GitHub repository, where all your projects should live for your lab.

Here is some readding and some exercises you can try at home, from our previous classes:

[Using the command-line](https://github.com/lifebit-ai/dry-bench-skills-for-researchers/blob/main/classes/class_1/1-using-the-command-line.ipynb)

[Why Git and GitHub?](https://github.com/lifebit-ai/dry-bench-skills-for-researchers/blob/main/classes/class_2/1-why-git-and-setup.md)

[Git Routines - Fork a Repository](https://github.com/lifebit-ai/dry-bench-skills-for-researchers/blob/main/classes/class_2/2-the-fork-git-routine.ipynb)

[Git Routines - Collaborating with Git](https://github.com/lifebit-ai/dry-bench-skills-for-researchers/blob/main/classes/class_2/3-the-add-push-git-routine.ipynb)

Extra reading for GitHub fun.
[Git GitHub Hello World 10 minute read](https://guides.github.com/activities/hello-world/)


## NOTE
An environment with all the required computational tools will be provided to the attendees of the workshop, hence there is no requirement for a personal computer with a specific OS (eg. Linux, Windows) for attending this workshop.

## Platform:
Learners will use Lifebit's CloudOS platform for this course. Requirements include the availability of a laptop, as they will then have the ability to log into a browser to access the course materials (Chrome preferred).
